src/replica/GetStatusQservCzarMgtRequest.cc

```diff
@@ -0,0 +1,60 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/GetStatusQservCzarMgtRequest.h"
+
+// LSST headers
+#include "lsst/log/Log.h"
+
+using namespace std;
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.GetStatusQservCzarMgtRequest");
+
+}  // namespace
+
+namespace lsst::qserv::replica {
+
+shared_ptr<GetStatusQservCzarMgtRequest> GetStatusQservCzarMgtRequest::create(
+        shared_ptr<ServiceProvider> const& serviceProvider, string const& czarName,
+        GetStatusQservCzarMgtRequest::CallbackType const& onFinish) {
+    return shared_ptr<GetStatusQservCzarMgtRequest>(
+            new GetStatusQservCzarMgtRequest(serviceProvider, czarName, onFinish));
+}
+
+GetStatusQservCzarMgtRequest::GetStatusQservCzarMgtRequest(
+        shared_ptr<ServiceProvider> const& serviceProvider, string const& czarName,
+        GetStatusQservCzarMgtRequest::CallbackType const& onFinish)
+        : QservCzarMgtRequest(serviceProvider, "QSERV_CZAR_GET_STATUS", czarName), _onFinish(onFinish) {}
+
+void GetStatusQservCzarMgtRequest::createHttpReqImpl(replica::Lock const& lock) {
+    string const service = "/status";
+    createHttpReq(lock, service);
+}
+
+void GetStatusQservCzarMgtRequest::notify(replica::Lock const& lock) {
+    LOGS(_log, LOG_LVL_TRACE, context() << __func__);
+    notifyDefaultImpl<GetStatusQservCzarMgtRequest>(lock, _onFinish);
+}
```

Comment:
That's correct. I too don't have a good solution for pushing this subclass-specific callback triggering into the base class. One alternative that I considered at some point was to use the following macros:
```
#define _REPLICA_NOTIFY_GENERATOR_DEF(CLASS) void notify(replica::Lock const& lock);

#define _REPLICA_NOTIFY_GENERATOR_IMPL(CLASS) \
void CLASS::notify(replica::Lock const& lock) { \
    LOGS(_log, LOG_LVL_TRACE, context() << __func__); \
    notifyDefaultImpl<CLASS>(lock, _onFinish); \
}
```
Such macro would be defined in the header file of the base class to be seen by subclasses The macros would allow injecting a single line into the header and implementation files of the relevant classes:
```
// File: replica/GetStatusQservCzarMgtRequest.h

class GetStatusQservCzarMgtRequest: public QservCzarMgtRequest {
protected:
    _REPLICA_NOTIFY_GENERATOR_DEF()
};
```
```
// File: replica/GetStatusQservCzarMgtRequest.cc

_REPLICA_NOTIFY_GENERATOR_IMPL(GetStatusQservCzarMgtRequest)
```
The second macro could be even simple (o mandatory class name parameter) should C++ have a predefined macro such as:
```
__class__
```
Or something like this. Unfortunately, no such macro is defined by the C++ standards or by GCC.
Another thing to worry about would be the contamination of headers with such macros and avoiding conflicts should two (or more) independent class hierarchies had to generate different code patterns.

Reply:
I'd really rather stay far away from macros. Aside from logging, they are tricky.